Upgrade Flask to v3 & commands to Click library

eos87 opened the pull request:
### Purpose
Migration to Flask v3 and commands to use `click` library. This PR does not go against `develop` branch but `async` instead so `develop` will continue working without disruption.

### What has changed
Most relevant ones:
- Commands have been migrated from `flask-script` to `click` + Flask command line management. Therefore the command runner is no longer `python manage.py <command>` but `flask newsroom <command>` name. At the same time, the commands name remain the same and the their parameters as well
- We're now targeting to minimum python v3.10, Mongo 6 and Elastic 7.17
- Using superdesk-core's async branch 
- Multiple dependencies have been upgraded because of their connection with Flask

This PR is connected with superdesk/superdesk-core#2632 although the one in superdesk is not ready to be merged. There are still multiple things to be checked and adjusted. As you may notice, we're depending on a "unstable" branch from superdesk that contains the most recent async changes and a partial upgrade to Flask v3. My idea is that we depend on that branch until Flask v3 migration is completed (at least stable enough) in superdesk 

Thanks for checking!

Resolves: [SDESK-7280] [SDESK-7281] 


[SDESK-7280]: https://sofab.atlassian.net/browse/SDESK-7280?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SDESK-7281]: https://sofab.atlassian.net/browse/SDESK-7281?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

eos87 committed:
Update flask v3 dependency & superdesk to flask v3

SDESK-7281

eos87 committed:
Upgrade newsroom to flask v3

SDESK-7281

eos87 committed:
Upgrade commands to use click library

SDESK-7280

eos87 committed:
Fix black and flake8 complains

MarkLark86 commented:
@eos87 Can you remove Python version `3.8` from all Github actions, as we're targeting a minimum of Python v3.10

eos87 committed:
Only target `3.10` on github actions

SDESK-7281

eos87 committed:
Update requirements with pip-compile

SDESK-7281

eos87 committed:
Fix linter complains and small timezone issue

SDESK-7281

eos87 committed:
Lock urllib3 version to avoid conflicts

SDESK-7281

eos87 committed:
Fix locked version

SDESK-7281

eos87 committed:
Downgrade urllib3 requirement

SDESK-7281

eos87 committed:
Bump python to 3.10 in github actions

SDESK-7281

eos87 committed:
Lock types-requests to also lock urllib3<2

SDESK-7281

eos87 committed:
Add async app to newsroom app

SDESK-7281

eos87 committed:
Update elastic and mongo in e2e services

SDESK-7281

MarkLark86 commented on .actions-docker-compose.yml:
```diff
@@ -1,19 +1,20 @@
 
     mongo:
-        image: mongo:3.6
+        image: mongo:4
```
Would be good to also update this to MongoDB 6 (see superdesk-core's [docker](https://github.com/superdesk/superdesk-core/blob/async/docker-compose.yml))

eos87 replied:
done

eos87 committed:
Avoid closing connection

Previously PyMongo client would auto-reconnect if the connection was closed. They have removed that in newer versions so we should not close the connection at this point.

SDESK-7281

eos87 committed:
Turn cursor into list as there is no count() method

SDESK-7281

eos87 committed:
Bump docker compose files to use mongo 6

SDESK-7281

eos87 commented on tests/core/test_oauth_clients.py:
```diff
@@ -1,10 +1,12 @@
 import base64
 
 
+@pytest.mark.skip(reason="Pending to figure out some issue with OAuth2")
```
I would like to have a separate ticket to take care of this part. Need to understand the whole logic to provide a proper fix.

MarkLark86 replied:
We have SDESK-7284 and NHUB-511 specifically to fix breaking changes. We can fix this as part of those tickets

eos87 committed:
Bump version and small fix

SDESK-7281

eos87 committed:
Merge branch 'async' into hg/flask-v3-unstable

eos87 committed:
Move template filter back to encoder class

SDESK-7281

eos87 committed:
Revert self.data.json_encoder_class

It was used after all

eos87 committed:
Back to encoder class

SDESK-7281

MarkLark86 commented on newsroom/commands/__init__.py:
```diff
@@ -1,24 +1,27 @@
-from .remove_expired import remove_expired  # noqa
-from .alerts import (  # noqa
+# disable "imported but unused" complains
+# flake8: noqa: F401
```
From what I read flake8 doesn't support, and don't want to support, file level rule exclusions.
Does this work, or does this disable flake8 for this entire file for all rules?

eos87 replied:
You're right. I just double checked it and indeed it disabled flake8 entirely for the whole file. Instead, what I did was to add a `per-file-ignores` in the `setup.cfg` because it really bothered me to see the #noqa so many times.

MarkLark86 commented on docker-compose.yml:
```diff
@@ -6,7 +6,7 @@ services:
 
     mongo:
-        image: mongo:3.6
+        image: mongo:6
```
Would be good to also update elastic to at least `7.17.22`, for this and e2e/docker-compose (that's what our test infra now uses). https://github.com/superdesk/superdesk-core/blob/async/docker-compose.yml#L9

MarkLark86 commented on requirements.txt:
```diff
@@ -1,5 +1,5 @@
 WTForms[email]>=3.1,<3.2
-flask>=1.1,<3.1
+flask>=3.0
```
We could probably remove some of the dependencies from this file (and leave it up to Superdesk-Core to define what versions are supported). Unless removing these has issues, I would recommend the following be removed:

* flask
* click
* flask-babel
* flask_pymongo
* eve-elastic
* MarkupSafe
* reportlab

eos87 replied:
Done. It worked well (all tests passed) I just needed to do a minor adjustment to Babel init process.

MarkLark86 reviewed (approved):
Looks good, just a couple small things

eos87 committed:
Add `per-file-ignores` instead of try file level disable

SDESK-7281

eos87 committed:
Update elasticsearch image in docker compose

SDESK-7281

eos87 committed:
Remove dependencies so pip figures out

SDESK-7281

eos87 committed:
Update babel setup
